File: forge-core/internal/kerneloperationalcontract/closure.go
package kerneloperationalcontract

import (
	"fmt"
	"sort"
)

func validateClosureCounts(value *KernelOperationalReferenceClosure) error {
	if value.Artifacts == nil || len(value.Artifacts) > maxArtifacts {
		return fmt.Errorf("artifacts cardinality must be 0..%d", maxArtifacts)
	}
	if value.ArtifactReceipts == nil || len(value.ArtifactReceipts) > maxArtifactReceipts {
		return fmt.Errorf("artifact_receipts cardinality must be 0..%d", maxArtifactReceipts)
	}
	if value.CapabilityInvocations == nil || len(value.CapabilityInvocations) < 1 ||
		len(value.CapabilityInvocations) > maxInvocations {
		return fmt.Errorf("capability_invocations cardinality must be 1..%d", maxInvocations)
	}
	if value.InteractionEvents == nil || len(value.InteractionEvents) > maxEvents {
		return fmt.Errorf("interaction_events cardinality must be 0..%d", maxEvents)
	}
	if value.ExecutionReceipts == nil || len(value.ExecutionReceipts) < 1 ||
		len(value.ExecutionReceipts) > maxExecutionReceipts {
		return fmt.Errorf("execution_receipts cardinality must be 1..%d", maxExecutionReceipts)
	}
	return nil
}

func validateClosureRecords(value *KernelOperationalReferenceClosure) error {
	if err := validateSortedUnique(value.Artifacts, "artifacts", maxArtifacts,
		false, validateArtifact); err != nil {
		return err
	}
	for index := range value.ArtifactReceipts {
		if err := validateArtifactReceipt(&value.ArtifactReceipts[index]); err != nil {
			return fmt.Errorf("artifact_receipts[%d]: %w", index, err)
		}
	}
	for index := range value.CapabilityInvocations {
		if err := validateInvocation(&value.CapabilityInvocations[index]); err != nil {
			return fmt.Errorf("capability_invocations[%d]: %w", index, err)
		}
	}
	for index := range value.InteractionEvents {
		if err := validateEvent(&value.InteractionEvents[index]); err != nil {
			return fmt.Errorf("interaction_events[%d]: %w", index, err)
		}
	}
	for index := range value.ExecutionReceipts {
		if err := validateExecutionReceipt(&value.ExecutionReceipts[index]); err != nil {
			return fmt.Errorf("execution_receipts[%d]: %w", index, err)
		}
	}
	return nil
}

func validateArtifactReceiptOrder(values []ArtifactReceipt) error {
	identities := make([]string, len(values))
	for index, value := range values {
		identities[index] = value.ArtifactReceiptID
	}
	if !sort.StringsAreSorted(identities) || hasDuplicate(identities) {
		return fmt.Errorf("artifact_receipts must be strictly identity-sorted and unique")
	}
	return nil
}

func validateClosureFields(value *KernelOperationalReferenceClosure, blank bool) error {
	if value == nil {
		return fmt.Errorf("KernelOperationalReferenceClosure is nil")
	}
	if err := validateRecordHeader(value.APIVersion, value.Kind, closureAPI, closureKind); err != nil {
		return err
	}
	if value.Canonicalization != canonicalization || value.Result != successMarker {
		return fmt.Errorf("closure canonicalization or result marker differs")
	}
	if err := validateIdentity(value.ClosureID, value.ClosureSHA256,
		closurePrefix, "closure", blank); err != nil {
		return err
	}
	if err := validateAttestations(value.Attestations); err != nil {
		return err
	}
	if err := validateClosureCounts(value); err != nil {
		return err
	}
	if err := validateClosureRecords(value); err != nil {
		return err
	}
	if err := validateArtifactReceiptOrder(value.ArtifactReceipts); err != nil {
		return err
	}
	return validateReferenceGraph(value)
}

func closureDigest(value *KernelOperationalReferenceClosure) (string, error) {
	blank, err := cloneValue(value)
	if err != nil {
		return "", err
	}
	blank.ClosureID, blank.ClosureSHA256 = "", ""
	if err := validateClosureFields(blank, true); err != nil {
		return "", err
	}
	return typedDigest(blank, closureDomain, maxClosureBytes)
}

// ValidateClosure validates every record, digest, projection, and supplied DAG edge.
func ValidateClosure(value *KernelOperationalReferenceClosure) error {
	if err := validateClosureFields(value, false); err != nil {
		return err
	}
	digest, err := closureDigest(value)
	if err != nil {
		return err
	}
	if value.ClosureSHA256 != digest {
		return fmt.Errorf("closure_sha256 does not match canonical preimage")
	}
	_, err = canonicalTyped(value, maxClosureBytes)
	return err
}

// SealClosure seals one exact blank-identity nonsemantic reference closure copy.
func SealClosure(value *KernelOperationalReferenceClosure) (*KernelOperationalReferenceClosure, error) {
	if value == nil || value.ClosureID != "" || value.ClosureSHA256 != "" {
		return nil, fmt.Errorf("sealing closure requires blank identity fields")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := closureDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.ClosureID, sealed.ClosureSHA256 = closurePrefix+digest, digest
	return sealed, ValidateClosure(sealed)
}

// DecodeClosure decodes one exact compact canonical closure with no trailing LF.
func DecodeClosure(data []byte) (*KernelOperationalReferenceClosure, error) {
	var value KernelOperationalReferenceClosure
	if err := decodeTypedExact(data, maxClosureBytes, &value); err != nil {
		return nil, err
	}
	return &value, ValidateClosure(&value)
}

File: forge-runtime/crates/domain/src/kernel_operational_contract/closure.rs
use sha2::{Digest, Sha256};

use super::{
    CANONICALIZATION, KernelOperationalContractError, KernelOperationalReferenceClosure,
    MAX_ARTIFACT_RECEIPTS, MAX_ARTIFACTS, MAX_CLOSURE_BYTES, MAX_EVENTS, MAX_EXECUTION_RECEIPTS,
    MAX_INVOCATIONS, SUCCESS_MARKER,
    constants::{CLOSURE_API, CLOSURE_DOMAIN, CLOSURE_KIND, CLOSURE_PREFIX},
    graph, invalid, primitives, records, wire,
};

pub(super) fn validate_counts(
    value: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    if value.artifacts.len() > MAX_ARTIFACTS
        || value.artifact_receipts.len() > MAX_ARTIFACT_RECEIPTS
        || !(1..=MAX_INVOCATIONS).contains(&value.capability_invocations.len())
        || value.interaction_events.len() > MAX_EVENTS
        || !(1..=MAX_EXECUTION_RECEIPTS).contains(&value.execution_receipts.len())
    {
        Err(invalid(
            "closure record cardinality is outside the frozen bounds",
        ))
    } else {
        Ok(())
    }
}

fn validate_receipt_order(
    value: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    if value
        .artifact_receipts
        .windows(2)
        .all(|pair| pair[0].artifact_receipt_id.as_bytes() < pair[1].artifact_receipt_id.as_bytes())
    {
        Ok(())
    } else {
        Err(invalid(
            "artifact_receipts must be strictly identity-sorted and unique",
        ))
    }
}

fn validate_body(
    value: &KernelOperationalReferenceClosure,
    allow_blank: bool,
) -> Result<(), KernelOperationalContractError> {
    if value.api_version != CLOSURE_API
        || value.kind != CLOSURE_KIND
        || value.canonicalization != CANONICALIZATION
        || value.result != SUCCESS_MARKER
    {
        return Err(invalid("closure constants or result marker differ"));
    }
    primitives::validate_identity_fields(
        &value.closure_id,
        &value.closure_sha256,
        CLOSURE_PREFIX,
        "closure",
        allow_blank,
    )?;
    primitives::validate_attestations(&value.attestations)?;
    validate_counts(value)?;
    primitives::validate_sorted_unique(
        &value.artifacts,
        "artifacts",
        MAX_ARTIFACTS,
        false,
        |artifact| primitives::validate_artifact(artifact, "artifacts"),
    )?;
    records::validate_records(
        &value.artifact_receipts,
        &value.capability_invocations,
        &value.interaction_events,
        &value.execution_receipts,
    )?;
    validate_receipt_order(value)?;
    graph::validate_reference_graph(value)
}

fn closure_digest(
    value: &KernelOperationalReferenceClosure,
) -> Result<String, KernelOperationalContractError> {
    let mut blank = value.clone();
    blank.closure_id.clear();
    blank.closure_sha256.clear();
    validate_body(&blank, true)?;
    let canonical = wire::canonical_typed_with_max(&blank, MAX_CLOSURE_BYTES)?;
    let mut hasher = Sha256::new();
    hasher.update(CLOSURE_DOMAIN);
    hasher.update(canonical.as_bytes());
    Ok(crate::governance_contract::codec::lower_hex(
        &hasher.finalize(),
    ))
}

/// Validates every record, digest, projection, and supplied DAG edge.
///
/// # Errors
///
/// Returns an error for any wire-independent semantic or self-seal violation.
pub fn validate_closure(
    value: &KernelOperationalReferenceClosure,
) -> Result<(), KernelOperationalContractError> {
    validate_body(value, false)?;
    if value.closure_sha256 != closure_digest(value)? {
        return Err(invalid(
            "closure_sha256 does not match the canonical preimage",
        ));
    }
    wire::canonical_typed_with_max(value, MAX_CLOSURE_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity nonsemantic reference closure copy.
///
/// # Errors
///
/// Returns an error for a nonblank identity or invalid closure, record, seal, or graph.
pub fn seal_closure(
    value: &KernelOperationalReferenceClosure,
) -> Result<KernelOperationalReferenceClosure, KernelOperationalContractError> {
    if !value.closure_id.is_empty() || !value.closure_sha256.is_empty() {
        return Err(invalid("sealing closure requires blank identity fields"));
    }
    let digest = closure_digest(value)?;
    let mut sealed = value.clone();
    sealed.closure_id = format!("{CLOSURE_PREFIX}{digest}");
    sealed.closure_sha256 = digest;
    validate_closure(&sealed)?;
    Ok(sealed)
}

/// Decodes one exact compact canonical closure with no trailing LF.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, or semantically invalid input.
pub fn decode_closure(
    bytes: &[u8],
) -> Result<KernelOperationalReferenceClosure, KernelOperationalContractError> {
    let value = wire::decode_typed(bytes, MAX_CLOSURE_BYTES)?;
    validate_closure(&value)?;
    Ok(value)
}
